allowContentType(s) expression docs #316 (#317)
enonic/doc-xp: docs/cms/input-types.adoc

  </config>
</input>
----
<1> *allowContentType* is used to limit the content types that may be selected using the <<allowContentType>>
<2> *allowPath* is used to limit the path of the content that may be selected using the <<allowPath Samples>>. By default, Content Selector
only displays content from the same site.
<3> *treeMode* toggles between tree and flat modes. Default value is `false` (flat mode). +
image:Content-selector-tree.png[Input box with a dropdown with a tree structure, 350]
<4> *hideToggleIcon* hides the toggle icon when set to `false`. +
image:Content-selector-no-icon.png[Input box without an icon, 500]

[#allowContentType]
=== Allow content type pattern

> There are two modes of parsing content type patterns in XP: `LEGACY` had been used before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` is current default mode. It can be switched back in <<../deployment/config.adoc#admin, config>>.

Reviewer: Explain `${app}` placeholder

> There are two modes of parsing content type patterns in XP: `LEGACY` had been used before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` is current default mode. It can be switched back in <<../deployment/config.adoc#admin, config>>.

Reviewer: explain `*` converted to `.*`

==== Legacy +

If pattern value can be found in any part of the content type name then it is accepted.

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if non of ':', '*' symbols or '${app}' template expression was used.

Reviewer: ```suggestion
Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols or '${app}' template expression was used.
```

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols or '${app}' template expression was used.

Reviewer: or starts with "`${app}`

==== Match +
A pattern will be treated as regexp and the whole content type name must match it to be accepted.

> It will be extended by the current application name if non of any special regex characters, ':' or '${app}' template expression was used.

Reviewer: list special regex characters `\^$.|?*+()[{`

> It will be extended by the current application name if none of any special regex characters, ':' or '${app}' template expression was used.

Reviewer: MATCH mode does not look for `${app}`, because it is sufficient to look for `$`

  </config>
</input>
----
<1> *allowContentType* is used to limit the content types that may be selected using the <<allowContentType>>
<2> *allowPath* is used to limit the path of the content that may be selected using the <<allowPath Samples>>. By default, Content Selector
only displays content from the same site.
<3> *treeMode* toggles between tree and flat modes. Default value is `false` (flat mode). +
image:Content-selector-tree.png[Input box with a dropdown with a tree structure, 350]
<4> *hideToggleIcon* hides the toggle icon when set to `false`. +
image:Content-selector-no-icon.png[Input box without an icon, 500]

[#allowContentType]
=== Allow content type pattern

> There are two modes of parsing content type patterns in XP: `LEGACY` had been used before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` is current default mode. It can be switched back in <<../deployment/config.adoc#admin, config>>.

Reviewer: ```suggestion
There are two modes of matching content type patterns in XP: `LEGACY` to simulate pattern matching before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` as current default mode. It can be configured in <<../deployment/config.adoc#admin, config>>.
```

> There are two modes of matching content type patterns in XP: `LEGACY` to simulate pattern matching before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` as current default mode. It can be configured in <<../deployment/config.adoc#admin, config>>.

Reviewer: there reason is a `imagesdir` at the top of the page. input-types.adoc has a different "default" imagesdir

==== Legacy +

If pattern value can be found in any part of the content type name then it is accepted.

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols was used or starts with '${app}' template expression.

Reviewer: ```suggestion
Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols wer used in pattern or pattern starts with '${app}'.
```

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols were used in pattern or pattern starts with '${app}'.

Reviewer: I accept your suggestion because it is more clear. it is slightly redundant, because `application-name:content-name` already contains `:`, thus first bullet is covered by second, but I guess it is fine

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols were used in pattern or pattern starts with '${app}'.

Reviewer: `+\{}[]()` don't' do anything special in LEGACY mode. This is to preserve backwards compatibility as much as possible

> Pattern will be extended by current application name if it's not set in <application-name:content-name> format or if none of ':', '*' symbols were used in pattern or pattern starts with '${app}'.

Reviewer 2: Nice. Is there any way it could _not_ be covered? Like if the app name contains `:` for some other reason? I'm guessing you're not allowed to use that character in app names, though 💁🏾

==== Match +

> A pattern will be treated as regexp and the whole content type name must match it to be accepted.

Reviewer: ```suggestion
A pattern will be treated as Java Regular Expression and the whole content type name must match it to be accepted.
```

> It will be extended by the current application name if none of any special regex characters `\^$.|?*+()[{` , ':' was used.

Reviewer: ```suggestion
It will be extended by the current application name if none of meta regular expression characters `\^$.|?*+()[{` , ':' were used.
```

==== Special handling

.'${app}' template expression
Most content type names start with the name of the application where they are stored. It may be quite messy and hard to maintain the whole application name in every pattern, so the '${app}' template expression can be used to replace it.

.'*' character

> There is a special handling of `*` symbol which differs from usual regex handling. It will be considered by parser as '0 or more of any character except linebreaks' meanwhile regex treat it as '1 or more of the preceding token'.

Reviewer: > Also, why is the * character treated differently?

I guess we have to keep it as is, It was implemented long ago and many applications are written with pure `*` in expressions. Issue to follow-up
https://github.com/enonic/xp/issues/8578

> There is a special handling of `*` symbol which differs from usual regex handling. It will be considered by parser as '0 or more of any character except linebreaks' meanwhile regex treat it as '1 or more of the preceding token'.

Reviewer 2: Cool, yeah. I thought about this a little bit last night, and it strikes me that there are a fair few issues you could run into that way. Would we consider changing it for XP 8? I don't know if it matters, but in the case that you do actually want the regular regex functionality of `*`, how would you do that? 

Anyway, back to the issue at hand: because `*` is treated differently, this should be clearly stated when the regexes are first brought up, I think.

> There is a special handling of `*` symbol which differs from usual regex handling. It will be considered by parser as '0 or more of any character except linebreaks' meanwhile regex treat it as '1 or more of the preceding token'.

Reviewer: Out of topic here, I guess. Discuss it in the issue.

I added clarification earlier in the text. Is it OK?

  </config>
</input>
----
<1> *allowContentType* is used to limit the content types that may be selected using the <<allowContentType>>
<2> *allowPath* is used to limit the path of the content that may be selected using the <<allowPath Samples>>. By default, Content Selector
only displays content from the same site.
<3> *treeMode* toggles between tree and flat modes. Default value is `false` (flat mode). +
image:Content-selector-tree.png[Input box with a dropdown with a tree structure, 350]
<4> *hideToggleIcon* hides the toggle icon when set to `false`. +
image:Content-selector-no-icon.png[Input box without an icon, 500]

[#allowContentType]
=== Allow content type pattern

> There are two modes of matching content type patterns in XP: `LEGACY` to simulate pattern matching before image:../images/xp-770.svg[XP 7.7.0,opts=inline] and `MATCH` as current default mode. It can be configured in <<../deployment/config.adoc#admin, config>>.
> 

Reviewer 2: I think this is a little unclear in how it's phrased. Could we clarify it a bit? Something like:

```suggestion
XP uses one of two modes when matching content type patterns:

`LEGACY`:: this mode simulates the pattern matching functionality that was the default before XP 7.7.0. image:../images/xp-770.svg[XP 7.7.0,opts=inline] 
`MATCH`:: this mode is the current default mode and has a few notable changes from `LEGACY` mode.

The modes are explained in further detail below. You can configure which mode to use in  <<../deployment/config.adoc#admin, the admin config>>.
```

And this might just be an error on my side, but when running the Asciidoc browser extension, it couldn't find the xp-770 svg. But this isn't run through any sort of build system or anything, so I don't know how it'll work on the doc page. I'm assuming you've got it under control, but just wanted to flag it 🙋🏾

==== Legacy +

If the pattern value can be found in any part of the content type name, then it is considered a match.

> The pattern will be prefixed by the current application name unless:
> - the pattern uses the <application-name:content-name> format

Reviewer 2: Sorry, I realized you need a newline here to make the list work as a list 😅 

```suggestion
The pattern will be prefixed by the current application name unless:

- the pattern uses the <application-name:content-name> format
```

- the pattern includes one of `:` or `*`
- the pattern starts with `${app}`.

==== Match +

> A pattern will be treated as Java Regular Expression and the whole content type name must match it to be accepted.

Reviewer: ```suggestion
A pattern will be treated as Java Regular Expression (with the exception to Special handling) and the whole content type name must match it to be accepted.
```

The pattern will be prefixed by the current application name if it contains none of the following characters `\^$.|?*+()[{:`.

==== Special handling

> .`${app}` template expression
> Most content type names start with the name of the application where they are stored. It may be quite messy and hard to maintain the whole application name in every pattern, so the '${app}' template expression can be used to replace it.
> 
> .'*' character

Reviewer 2: Just further to the conversation we had previously regarding quotes: If possible, I think we should be consistent in the uses of quotes. These lines and the next one mix single quotes and backticks. As before, the choice between single quotes or backticks is up to you, but I suggest something like the following:

```suggestion
.`${app}` template expression
Most content type names start with the name of the application where they are stored. It may be quite messy and hard to maintain the whole application name in every pattern, so the `${app}` template expression can be used to replace it.

.`*` character
```

`MATCH`:: this mode is the current default mode and has a few notable changes from `LEGACY` mode.

The modes are explained in further detail below. You can configure which mode to use in  <<../deployment/config.adoc#admin, the admin config>>.
==== Legacy +

If the pattern value can be found in any part of the content type name, then it is considered a match.
The pattern will be prefixed by the current application name unless:

- the pattern uses the <application-name:content-name> format
- the pattern includes one of `:` or `*`
- the pattern starts with `${app}`.

==== Match +
A pattern will be treated as Java Regular Expression (with the exception to Special handling) and the whole content type name must match it to be accepted.
The pattern will be prefixed by the current application name if it contains none of the following characters `\^$.|?*+()[{:`.

==== Special handling

.`${app}` template expression
Most content type names start with the name of the application where they are stored. It may be quite messy and hard to maintain the whole application name in every pattern, so the `${app}` template expression can be used to replace it.

> .`*` character
> There is a special handling of `*` symbol which differs from usual regex handling. It will be considered by parser as '0 or more of any character except linebreaks' meanwhile regex treat it as '1 or more of the preceding token'.

Reviewer 2: I wonder whether clarifying this with some examples would be good? Something like this? 🤓 

```suggestion
.The '*' character
In most regex engines, the `*` character means "0 or more of the preceding token". However, XP treats it differently and interprets it as "0 or more of any character(s) except for line breaks". In that way, it's similar to https://en.wikipedia.org/wiki/Glob_%28programming%29[glob patterns].

To further clarify, in XP:

- `*day` would match "monday", "day", and "mayday". This pattern would be invalid in most regex engines.
- `p*age` would match "page" and "package", but _not_ "age".
- `monk*` would match "monkey" and  "monk", but _not_ "mon".
```

Also, `*` usually means 0 or more of the preceding token, not 1 or more (at least in my experience).

<3> *treeMode* toggles between tree and flat modes. Default value is `false` (flat mode). +
image:Content-selector-tree.png[Input box with a dropdown with a tree structure, 350]
<4> *hideToggleIcon* hides the toggle icon when set to `false`. +
image:Content-selector-no-icon.png[Input box without an icon, 500]

[#allowContentType]
=== Allow content type pattern
XP uses one of two modes when matching content type patterns:

`LEGACY`:: this mode simulates the pattern matching functionality that was the default before XP 7.7.0. image:../images/xp-770.svg[XP 7.7.0,opts=inline] 
`MATCH`:: this mode is the current default mode and has a few notable changes from `LEGACY` mode.

The modes are explained in further detail below. You can configure which mode to use in  <<../deployment/config.adoc#admin, the admin config>>.
==== Legacy +

If the pattern value can be found in any part of the content type name, then it is considered a match.
The pattern will be prefixed by the current application name unless:

- the pattern uses the <application-name:content-name> format
- the pattern includes one of `:` or `*`
- the pattern starts with `${app}`.

==== Match +

> A pattern will be treated as Java Regular Expression (with the exception to Special handling) and the whole content type name must match it to be accepted.
> The pattern will be prefixed by the current application name if it contains none of the following characters `\^$.|?*+()[{:`.
> 
> ==== Special handling

Reviewer 2: My final suggestion 🙋🏾  I mentioned this briefly in a previous thread on this PR, but thought I'd put in a suggestion for it. This adds a link to the section about XP's differences and renames the title slightly.

I also reworded the sentence about characters that stop a pattern from being prefixed with the current application name for clarity.

```suggestion
The pattern will be treated as a Java Regular Expression (with <<#regex-differences, a few notable exceptions>>) and the whole content type name must match it to be accepted.
The pattern will be prefixed by the current application name if it does _not_ contain any of the following characters: `\^$.|?*+()[{:`.

[#regex-differences]
==== XP patterns: special cases

XP's patterns have a few cases where they differ from normal Java Regular Expression syntax:
```

All content types starting with banana (using regular expressions) +

>   <allowContentType>*:banana*</allowContentType>

Reviewer: it was a certain documentation issue `*:^banana` - it never worked.
So fix the docs a little here.

You have the right point, we have a deviation from correct Java RegEx with `*` but we can do nothing about it for backwards compatibility.

>   <allowContentType>*:banana*</allowContentType>

Reviewer 2: Good catch on removing the caret, then! 😄 

But here too, we might want to reference back to the section on `*` being treated differently in XP. What do you think?

>   <allowContentType>*:banana*</allowContentType>

Reviewer: Any suggestion how we do it in examples without too much bloat?

>   <allowContentType>*:banana*</allowContentType>

Reviewer 2: I've had a look at it on the rendered page asciidoc output now. When doing that, the relation to this section is much clearer, and I don't think we need to reiterate the special `*` handling:

![image](https://user-images.githubusercontent.com/17786332/124434120-ca4db580-dd73-11eb-8a1f-b90f5d776a08.png)

So I was probably a bit too eager here. If you think it's fine (_I_ do!), then please feel free to resolve this ☺️
